File: apps/TiflisCodeAndroid/app/src/main/java/io/tiflis/code/data/websocket/CommandSender.kt
/*
 * Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
 * Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
 */

package io.tiflis.code.data.websocket

import android.util.Log
import io.tiflis.code.domain.models.ConnectionState
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min
import kotlin.math.pow

/**
 * Centralized service for safe command sending with atomic auth checking,
 * retry logic, and command queuing during disconnections.
 *
 * Key guarantees:
 * - Auth state is checked immediately before send (no race condition)
 * - Failed commands are retried with exponential backoff
 * - Commands are queued during brief disconnections
 * - No user command is silently dropped
 */
@Singleton
class CommandSender @Inject constructor(
    private val webSocketClient: WebSocketClient
) {
    companion object {
        private const val TAG = "CommandSender"

        /** Maximum queue size to prevent memory issues */
        private const val MAX_QUEUE_SIZE = 50

        /** Maximum time a command can stay queued (60 seconds) */
        private const val MAX_QUEUE_TIME_MS = 60_000L

        /** Base delay for exponential backoff (0.5 seconds) */
        private const val BASE_RETRY_DELAY_MS = 500L

        /** Maximum retry delay cap (4 seconds) */
        private const val MAX_RETRY_DELAY_MS = 4_000L
    }

    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    /** Queue for commands waiting to be sent during disconnection */
    private val commandQueue = mutableListOf<QueuedCommand>()

    /** Lock for thread-safe queue access */
    private val queueLock = Any()

    init {
        // Observe connection state for queue processing
        scope.launch {
            webSocketClient.connectionState.collect { state ->
                // Process queue when connection is established (Connected, Verified, or Degraded)
                if (state.isConnected) {
                    processQueue()
                }
            }
        }
    }

    /** Number of commands currently in the queue */
    val pendingCommandCount: Int
        get() = synchronized(queueLock) { commandQueue.size }

    /**
     * Sends a command with atomic auth checking, retry logic, and optional queuing.
     *
     * @param config Command configuration
     * @return Result indicating success, failure, or queued status
     */
    suspend fun send(config: CommandConfig): CommandSendResult {
        val commandName = config.debugName ?: config.type

        // 1. Check connection state ATOMICALLY before send
        val currentState = webSocketClient.connectionState.value

        if (!currentState.isConnected) {
            // Not connected - queue if allowed
            Log.d(TAG, "Not connected for $commandName, state: $currentState")
            if (config.shouldQueue) {
                return queueCommand(config)
            }
            return CommandSendResult.Failure(CommandSendError.NotAuthenticated)
        }

        // 2. Attempt to send with retry logic
        return attemptSend(config, attempt = 0)
    }

    /**
     * Cancels any pending queued commands for a specific session.
     *
     * @param sessionId The session ID to cancel commands for
     */
    fun cancelPendingCommands(sessionId: String) {
        synchronized(queueLock) {
            val beforeCount = commandQueue.size
            commandQueue.removeAll { command ->
                val msgSessionId = extractSessionId(command.config.message)
                msgSessionId == sessionId
            }
            val removedCount = beforeCount - commandQueue.size
            if (removedCount > 0) {
                Log.d(TAG, "Cancelled $removedCount pending commands for session $sessionId")
            }
        }
    }

    /**
     * Cancels all pending queued commands.
     */
    fun cancelAllPendingCommands() {
        synchronized(queueLock) {
            val count = commandQueue.size
            commandQueue.clear()
            if (count > 0) {
                Log.d(TAG, "Cancelled all $count pending commands")
            }
        }
    }

    private suspend fun attemptSend(config: CommandConfig, attempt: Int): CommandSendResult {
        val commandName = config.debugName ?: config.type

        // CRITICAL: Re-check auth state immediately before send
        // This closes the race condition window
        val currentState = webSocketClient.connectionState.value
        if (!currentState.isConnected) {
            Log.w(TAG, "Auth state changed before send for $commandName")
            if (config.shouldQueue) {
                return queueCommand(config)
            }
            return CommandSendResult.Failure(CommandSendError.NotAuthenticated)
        }

        return try {
            val success = webSocketClient.sendMessage(config.message)
            if (success) {
                if (attempt > 0) {
                    Log.d(TAG, "$commandName sent successfully after $attempt retries")
                } else {
                    Log.d(TAG, "$commandName sent successfully")
                }
                CommandSendResult.Success
            } else {
                handleSendFailure(config, attempt, Exception("WebSocket send returned false"))
            }
        } catch (e: Exception) {
            handleSendFailure(config, attempt, e)
        }
    }

    private suspend fun handleSendFailure(
        config: CommandConfig,
        attempt: Int,
        error: Exception
    ): CommandSendResult {
        val commandName = config.debugName ?: config.type
        Log.w(TAG, "$commandName send failed (attempt ${attempt + 1}/${config.maxRetries + 1}): ${error.message}")

        // Check if we should retry
        if (attempt < config.maxRetries) {
            // Calculate backoff delay with exponential increase
            val delay = calculateBackoff(attempt)
            Log.d(TAG, "Retrying $commandName in ${delay}ms...")

            // Wait for backoff
            delay(delay)

            // Retry (recursive call will re-check auth)
            return attemptSend(config, attempt + 1)
        }

        // Max retries exceeded - queue if allowed, otherwise fail
        if (config.shouldQueue) {
            Log.d(TAG, "Max retries exceeded for $commandName, queuing")
            return queueCommand(config)
        }

        return CommandSendResult.Failure(CommandSendError.MaxRetriesExceeded)
    }

    private fun calculateBackoff(attempt: Int): Long {
        // Exponential backoff: 0.5s, 1s, 2s, 4s (capped)
        val delay = BASE_RETRY_DELAY_MS * 2.0.pow(attempt.toDouble())
        return min(delay.toLong(), MAX_RETRY_DELAY_MS)
    }

    private fun queueCommand(config: CommandConfig): CommandSendResult {
        val commandName = config.debugName ?: config.type

        synchronized(queueLock) {
            // Enforce queue size limit - remove oldest if full
            if (commandQueue.size >= MAX_QUEUE_SIZE) {
                val removed = commandQueue.removeAt(0)
                Log.w(TAG, "Queue full, dropped oldest command: ${removed.config.debugName ?: removed.config.type}")
            }

            val queuedCommand = QueuedCommand(
                config = config,
                queuedAt = System.currentTimeMillis()
            )
            commandQueue.add(queuedCommand)

            Log.d(TAG, "Queued $commandName (queue size: ${commandQueue.size})")
        }

        return CommandSendResult.Queued
    }

    private fun processQueue() {
        scope.launch {
            val commandsToProcess: List<QueuedCommand>

            synchronized(queueLock) {
                val now = System.currentTimeMillis()

                // Filter out expired commands
                val expiredCount = commandQueue.count { now - it.queuedAt > MAX_QUEUE_TIME_MS }
                if (expiredCount > 0) {
                    Log.d(TAG, "Removing $expiredCount expired commands from queue")
                }
                commandQueue.removeAll { command ->
                    now - command.queuedAt > MAX_QUEUE_TIME_MS
                }

                // Take all remaining commands
                commandsToProcess = commandQueue.toList()
                commandQueue.clear()
            }

            if (commandsToProcess.isEmpty()) return@launch

            Log.d(TAG, "Processing ${commandsToProcess.size} queued commands")

            for ((index, command) in commandsToProcess.withIndex()) {
                // Small delay between queued commands to avoid overwhelming the server
                if (index > 0) {
                    delay(100)
                }

                // Re-send (creates new CommandConfig without re-queuing on failure)
                val noQueueConfig = command.config.copy(shouldQueue = false)
                val result = attemptSend(noQueueConfig, attempt = 0)

                if (result is CommandSendResult.Failure) {
                    Log.w(TAG, "Queued command ${command.config.debugName ?: command.config.type} failed: ${result.error.message}")
                }
            }

            Log.d(TAG, "Finished processing queued commands")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun extractSessionId(message: Map<String, Any?>): String? {
        // Try direct session_id
        val directId = message["session_id"] as? String
        if (directId != null) return directId

        // Try in payload
        val payload = message["payload"] as? Map<String, Any?>
        return payload?.get("session_id") as? String
    }
}

/**
 * A command waiting in the queue.
 */
private data class QueuedCommand(
    val config: CommandConfig,
    val queuedAt: Long
)

File: apps/TiflisCode/Shared/Services/WebSocket/CommandSender.swift
//
//  CommandSender.swift
//  TiflisCode
//
//  Copyright (c) 2025 Roman Barinov <rbarinov@gmail.com>
//  Licensed under the Functional Source License (FSL-1.1-NC). See LICENSE file for details.
//

import Foundation
import Combine

/// Centralized service for safe command sending with atomic auth checking,
/// retry logic, and command queuing during disconnections.
///
/// Key guarantees:
/// - Auth state is checked immediately before send (no race condition)
/// - Failed commands are retried with exponential backoff
/// - Commands are queued during brief disconnections
/// - No user command is silently dropped
@MainActor
final class CommandSender: CommandSending {
    // MARK: - Dependencies

    private weak var connectionService: ConnectionServicing?
    private let webSocketClient: WebSocketClientProtocol

    // MARK: - Queue State

    /// Queue for commands waiting to be sent during disconnection
    private var commandQueue: [QueuedCommand] = []

    /// Maximum queue size to prevent memory issues
    private let maxQueueSize = 50

    /// Maximum time a command can stay queued (60 seconds)
    private let maxQueueTime: TimeInterval = 60.0

    // MARK: - Retry Configuration

    /// Base delay for exponential backoff (0.5 seconds)
    private let baseRetryDelay: TimeInterval = 0.5

    /// Maximum retry delay cap (4 seconds)
    private let maxRetryDelay: TimeInterval = 4.0

    // MARK: - Combine

    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init(connectionService: ConnectionServicing, webSocketClient: WebSocketClientProtocol) {
        self.connectionService = connectionService
        self.webSocketClient = webSocketClient

        observeConnectionState()
    }

    // MARK: - Public API

    var pendingCommandCount: Int {
        commandQueue.count
    }

    func send(_ config: CommandConfig) async -> CommandSendResult {
        let commandName = config.debugName ?? "unknown"

        // 1. Check authentication state ATOMICALLY before send
        guard let service = connectionService else {
            print("⚠️ CommandSender: No connection service for \(commandName)")
            return .failure(.connectionLost)
        }

        let authState = service.connectionState

        switch authState {
        case .authenticated, .verified:
            // Good to send
            break

        case .degraded:
            // Connection is degraded but may still work - try to send
            print("⚠️ CommandSender: Connection degraded, attempting \(commandName) anyway")

        case .connecting, .connected, .authenticating:
            // Not yet authenticated - queue if allowed
            print("⏳ CommandSender: Not yet authenticated for \(commandName), state: \(authState)")
            if config.shouldQueue {
                return queueCommand(config)
            }
            return .failure(.notAuthenticated)

        case .disconnected, .error:
            // Disconnected - queue if allowed
            print("🔴 CommandSender: Disconnected for \(commandName), state: \(authState)")
            if config.shouldQueue {
                return queueCommand(config)
            }
            return .failure(.notAuthenticated)
        }

        // 2. Attempt to send with retry logic
        return await attemptSend(config, attempt: 0)
    }

    func sendThrowing(_ config: CommandConfig) async throws {
        let result = await send(config)
        switch result {
        case .success:
            return
        case .queued:
            // Queued is acceptable for throwing variant - command will be sent later
            return
        case .failure(let error):
            throw error
        }
    }

    func cancelPendingCommands(for sessionId: String) {
        let beforeCount = commandQueue.count
        commandQueue.removeAll { command in
            guard let msgSessionId = extractSessionId(from: command.config.message) else {
                return false
            }
            return msgSessionId == sessionId
        }
        let removedCount = beforeCount - commandQueue.count
        if removedCount > 0 {
            print("🗑️ CommandSender: Cancelled \(removedCount) pending commands for session \(sessionId)")
        }
    }

    func cancelAllPendingCommands() {
        let count = commandQueue.count
        commandQueue.removeAll()
        if count > 0 {
            print("🗑️ CommandSender: Cancelled all \(count) pending commands")
        }
    }

    // MARK: - Private Methods

    private func observeConnectionState() {
        connectionService?.connectionStatePublisher
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                guard let self = self else { return }

                switch state {
                case .authenticated, .verified:
                    // Connection restored - process queued commands
                    if !self.commandQueue.isEmpty {
                        print("✅ CommandSender: Connection restored, processing \(self.commandQueue.count) queued commands")
                        Task { @MainActor in
                            await self.processQueue()
                        }
                    }
                default:
                    break
                }
            }
            .store(in: &cancellables)
    }

    private func attemptSend(_ config: CommandConfig, attempt: Int) async -> CommandSendResult {
        let commandName = config.debugName ?? "unknown"

        // CRITICAL: Re-check auth state immediately before send
        // This closes the race condition window
        guard let service = connectionService else {
            return .failure(.connectionLost)
        }

        let currentState = service.connectionState
        guard currentState.isConnected else {
            print("⚠️ CommandSender: Auth state changed before send for \(commandName)")
            if config.shouldQueue {
                return queueCommand(config)
            }
            return .failure(.notAuthenticated)
        }

        do {
            try webSocketClient.sendMessage(config.message)
            if attempt > 0 {
                print("✅ CommandSender: \(commandName) sent successfully after \(attempt) retries")
            } else {
                print("📤 CommandSender: \(commandName) sent successfully")
            }
            return .success
        } catch {
            print("❌ CommandSender: \(commandName) send failed (attempt \(attempt + 1)/\(config.maxRetries + 1)): \(error)")

            // Check if we should retry
            if attempt < config.maxRetries {
                // Calculate backoff delay with exponential increase
                let delay = calculateBackoff(attempt: attempt)
                print("⏱️ CommandSender: Retrying \(commandName) in \(String(format: "%.1f", delay))s...")

                // Wait for backoff
                try? await Task.sleep(for: .seconds(delay))

                // Retry (recursive call will re-check auth)
                return await attemptSend(config, attempt: attempt + 1)
            }

            // Max retries exceeded - queue if allowed, otherwise fail
            if config.shouldQueue {
                print("📦 CommandSender: Max retries exceeded for \(commandName), queuing")
                return queueCommand(config)
            }

            return .failure(.maxRetriesExceeded)
        }
    }

    private func calculateBackoff(attempt: Int) -> TimeInterval {
        // Exponential backoff: 0.5s, 1s, 2s, 4s (capped)
        let delay = baseRetryDelay * pow(2.0, Double(attempt))
        return min(delay, maxRetryDelay)
    }

    private func queueCommand(_ config: CommandConfig) -> CommandSendResult {
        let commandName = config.debugName ?? "unknown"

        // Enforce queue size limit - remove oldest if full
        if commandQueue.count >= maxQueueSize {
            let removed = commandQueue.removeFirst()
            print("⚠️ CommandSender: Queue full, dropped oldest command: \(removed.config.debugName ?? "unknown")")
        }

        let queuedCommand = QueuedCommand(
            config: config,
            queuedAt: Date()
        )
        commandQueue.append(queuedCommand)

        print("📦 CommandSender: Queued \(commandName) (queue size: \(commandQueue.count))")
        return .queued
    }

    private func processQueue() async {
        let now = Date()

        // Filter out expired commands
        let expiredCount = commandQueue.filter { now.timeIntervalSince($0.queuedAt) > maxQueueTime }.count
        if expiredCount > 0 {
            print("⏰ CommandSender: Removing \(expiredCount) expired commands from queue")
        }
        commandQueue.removeAll { command in
            now.timeIntervalSince(command.queuedAt) > maxQueueTime
        }

        // Process remaining commands
        let commandsToProcess = commandQueue
        commandQueue.removeAll()

        print("🔄 CommandSender: Processing \(commandsToProcess.count) queued commands")

        for (index, command) in commandsToProcess.enumerated() {
            // Small delay between queued commands to avoid overwhelming the server
            if index > 0 {
                try? await Task.sleep(for: .milliseconds(100))
            }

            // Re-send (creates new CommandConfig without re-queuing on failure)
            let noQueueConfig = CommandConfig(
                message: command.config.message,
                maxRetries: command.config.maxRetries,
                shouldQueue: false, // Don't re-queue on failure during queue processing
                debugName: command.config.debugName
            )

            let result = await attemptSend(noQueueConfig, attempt: 0)
            if case .failure(let error) = result {
                print("❌ CommandSender: Queued command \(command.config.debugName ?? "unknown") failed: \(error)")
            }
        }

        print("✅ CommandSender: Finished processing queued commands")
    }

    private func extractSessionId(from message: [String: Any]) -> String? {
        // Try direct session_id
        if let sessionId = message["session_id"] as? String {
            return sessionId
        }
        // Try in payload
        if let payload = message["payload"] as? [String: Any],
           let sessionId = payload["session_id"] as? String {
            return sessionId
        }
        return nil
    }
}

// MARK: - Supporting Types

private struct QueuedCommand {
    let config: CommandConfig
    let queuedAt: Date
}
